llama_index/node_parser/code_hierarchy.py

```diff
@@ -0,0 +1,284 @@
+
+from typing import Any, List, Optional, Sequence
+
+from pydantic import Field
+from tree_sitter import Node
+from llama_index.callbacks.base import CallbackManager
+from llama_index.callbacks.schema import CBEventType, EventPayload
+from llama_index.node_parser.extractors.metadata_extractors import MetadataExtractor
+from llama_index.node_parser.interface import NodeParser
+from llama_index.node_parser.simple import SimpleNodeParser
+from llama_index.schema import BaseNode, Document, NodeRelationship, TextNode
+from llama_index.text_splitter.code_splitter import CodeSplitter
+from llama_index.utils import get_tqdm_iterable
+from pydantic import BaseModel
+
+# TODO: Expand these for new languages
+DEFAULT_NAME_IDENTIFIERS = {
+    'python': ["identifier"],
+    'javascript': ["identifier", "type_identifier", "property_identifier"],
+    'typescript': ["identifier", "type_identifier", "property_identifier"],
+    'html': ["start_tag"],
+    'cpp': ["function_declarator", "type_identifier"]
+}
+
+DEFAULT_SPLIT_ON_TYPES = {
+    'python': ["function_definition", "class_definition"],
+    'javascript': ["function_declaration", "class_declaration", "lexical_declaration", "method_definition"],
+    'typescript': ["function_declaration", "class_declaration", "interface_declaration", "method_definition", "lexical_declaration"],
+    'html': ["element"],
+    'cpp': ["class_specifier", "function_definition"]
+}
+
+class _ScopeItem(BaseModel):
+    """Like a Node from tree_sitter, but with only the str information we need."""
+    name: str
+    type: str
+
+class _ChunkNodeOutput(BaseModel):
+    """The output of a chunk_node call."""
+    this_document: Optional[TextNode]
+    children_documents: List[TextNode]
+
+class CodeBlockNodeParser(NodeParser):
+    """Split code using a AST parser.
+
+    Add metadata about the scope of the code block and relationships between
+    code blocks.
+    """
+
+    @classmethod
+    def class_name(cls) -> str:
+        """Get class name."""
+        return cls.__name__
+
+    language: str = Field(
+        description="The programming languge of the code being split."
+    )
+    split_on_types: List[str] = Field(
+        description="The types of nodes to split on."
+    )
+    name_identifiers: List[str] = Field(
+        description="The types of nodes to use as the name of the scope."
+    )
+    min_characters: int = Field(
+        default=0, description="Minimum number of characters per chunk."
```

Comment:
Does having a chunk with zero characters make sense?

Reply:
It just means it's not limiting the size, I didnt want to add an opinion on it to the defaults but I do think it's most useful when at least set above 2. In python I'd probably set it to the line length of 180 or something.

Reply:
Ah I see, makes sense!